### PR TITLE
Federated DR functionality

### PR DESCRIPTION
**Pull request by kaveenh**

### CODE REVIEW

File: examples/data_readiness/grpc/resources/configs/mnist/client_1.yaml
Comment (Zilinghan) [inline examples/data_readiness/grpc/resources/configs/mnist/client_1.yaml:1]: As `dr_metrics` is the same for all clients in most cases, it should be put into server configuration file under `client_configs`

File: examples/data_readiness/globus_compute/resources/config_gc/clients.yaml
Comment (Zilinghan) [inline examples/data_readiness/globus_compute/resources/config_gc/clients.yaml:1]: Similarly, the `dr_metrics` should be in the server configuration file, as it is a general client configuration

**Comment by Zilinghan** [comment]
Hi @kaveenh, thanks a lot on your great effort! This PR looks great to me overall, and I just left few minor points regrading the code structure and function/file naming points above.

For the examples, I saw you are creating a new folder `examples/dr_metrics`. However, most of the files within that folder is a bit duplicated compared with the `examples` folder, so I suggest the following way to restructure the dr metrics example in the following way
- Put all dr related metrics directly into the corresponding configuration folder, e.g., for `examples/resources/configs/mnist/server_fedavg.yaml`, just make it having the the following content with an additional flag parameter `generate_dr_report`.
    ```
    data_readiness_configs:
      generate_dr_report: True
      output_dirname: "./output"
      output_filename: "data_readiness_report"
    ```
- Also, as indicated in the above comments, the `dr_metrics`, which is a general client configuration that should be shared among all clients, putting it into server configuration file's `client_configs` simplifies the setting up.
- To run with dr metric, the runners (e.g., `run_mpi.py`, `run_server.py`, etc.) may have some changes, just put the changes into the corresponding runners within the `examples` folder.

**Review by Zilinghan — CHANGES_REQUESTED** [review]

**Comment by kaveenh** [comment]
Hi Zilinghan,

Thank you for the valuable feedback. I will revise the proposed changes and resubmit.

Best regards
Kaveen
________________________________

File: src/appfl/agent/dr_agent.py
[the remark this answers is not in the document]
Comment (Zilinghan) [reply src/appfl/agent/dr_agent.py:1]: As this dr_agent is basically used by the client side to generate DR report, I recommend to put its functionalities within the `ClientAgent` directly

**Comment by Zilinghan** [comment]
@kaveenh Could you please also help to revise this tutorial accordingly for your experience when adding your more advanced DR usecase: https://github.com/APPFL/APPFL/blob/main/docs/tutorials/examples_custom_action.rst (mine version is too simplified

**Comment by kaveenh** [comment]
In the meantime, I will work on the documentation with the DR use cases.

**Comment by Zilinghan** [comment]
Hi @kaveenh, I fixed some minor points for this PR in [this commit](https://github.com/APPFL/APPFL/pull/202/commits/4c8456796d58fe3fffbced0398f3b68ccaf44b2a), and now it is ready for merging once the documentation is uploaded. Please note the following key changes in my fix:

- As the documentation mentions that user runs all scripts within the `examples` folder, such as `python globus_compute/run.py`, instead of `python run.py`, I changed all `../resources/` to `./resources/` in the configuration files.
- I updated the `appfl.comm.grpc.grpc_server_communicator` and `appfl.comm.mpi.mpi_server_communicator` to make the DR action a synchronous action, i.e., return to clients once receiving requests from all clients - Please include this change into the documentation.

Again, thanks a lot for your great work!

**Comment by kaveenh** [comment]
Hi @Zilinghan 

Thank you for the valuable feedback and the help in reviewing the PR. I will note the changes and include them in the documentation.

**Comment by kaveenh** [comment]
Hi @Zilinghan,

The documentation is now updated [here](https://github.com/APPFL/APPFL/pull/202/commits/18422e43574a1984700f7765b4e018fcb9b2d246). Please let me know if you have any concerns or suggestions for changes.

Thank you

**Comment by Zilinghan** [comment]
> Hi @Zilinghan,
> 
> The documentation is now updated [here](https://github.com/APPFL/APPFL/pull/202/commits/18422e43574a1984700f7765b4e018fcb9b2d246). Please let me know if you have any concerns or suggestions for changes.
> 
> Thank you

@kaveenh Thanks, the documentation looks perfect to me. I will go ahead to approve and merge this PR.

**Review by Zilinghan — APPROVED** [review]

**Pull request merged**